File: 90.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        dfs(res, new ArrayList<Integer>(), nums, 0);
        return res;
    }
    private void dfs(List<List<Integer>> res, List<Integer> tmp, int[] nums, int start){
        res.add(new ArrayList<>(tmp));
        for(int i=start;i<nums.length;i++){
            if(i>start&&nums[i]==nums[i-1]) continue;
            tmp.add(nums[i]);
            dfs(res,tmp,nums,i+1);
            tmp.remove(tmp.size()-1);
        }
    }
}

File: 90.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> t;
        res.push_back(t);
        int size;
        int start,end;
        for(int index=0;index<nums.size();index++){
            start = 0;
            if(index>0&&nums[index]==nums[index-1])
                start = size;
            size = res.size();
            end = size;
            for(int i=start;i<end;i++){
                vector<int> temp(res[i]);
                temp.push_back(nums[index]);
                res.push_back(temp);
            }
        }
        return res;
    }
};